drivers: modbus_rtu: add initial support (v2)

basilfx opened the pull request:
### Contribution description

This PR is a reboot of https://github.com/RIOT-OS/RIOT/pull/13913 that got stale. I cloned it, rebased it onto master, adapted it to latest standards (hopefully) and proposed https://github.com/RIOT-OS/RIOT/pull/19819. But after some actual testing, I decided that both PRs were not really useful, and made a lot of assumptions. For more flexibility, I rewrote most of the code. Some notable improvements:

* No more hard-coded addresses. If you request address 1000, the slave does not need to allocate a buffer to hold 1000 registers.
* Slave implementation uses a request callback, which makes it much simpler to implement dynamic behavior.
* Support for exceptions.
* Better documented, consistent naming with Modbus specification.

Contrary to the original PR, I removed a lot of `#ifdefs` that limited which Modbus functions would be supported. It was hard to manage IMHO.

I also extended the test application to test using a single or multiple devices. An example application that provides a Modbus slave implementation is also provided.

### Testing procedure
A test application is provided in `tests/drivers/modbus`.

There are multiple ways to test this. The easiest is to connect RX/TX of one UART to TX/RX of another UART. That way, Modbus runs over RS-232, which should be sufficient to test majority of the implementation. If you take a board with two additional UARTs, then you can use the loopback mode, which will perform additional assertions by comparing data directly.

Alternatively, connect two or more boards via a RS-485 network. Let one run as master and the other as slaves.

See [here](https://github.com/basilfx/RIOT/tree/feature/modbus_rtu_v2/tests/drivers/modbus#modbus-driver-test) for more information.

<img src="https://github.com/RIOT-OS/RIOT/assets/815976/241bdc85-969c-4ac4-9833-f146e0fb8ef9" width="256">

### Issues/PRs references
Fixes #13869.

kfessel commented on drivers/include/modbus_rtu.h:
```diff
@@ -0,0 +1,193 @@
+ * @brief   Modbus RTU device parameters
+ */
+typedef struct {
+    /**
+     * @brief   RTS pin. @p GPIO_UNDEF if not used.
+     */
+    gpio_t pin_rts;
```
why dont you use ?
```suggestion
    gpio_t pin_rts; /**<<  RTS pin. @p GPIO_UNDEF if not used.*/
```

basilfx replied:
I will update this. I just re-used what was already provided in the first PR :-)

basilfx replied:
I have addressed this for `modbus_rtu_params_t` and `modbus_rtu_t`. I did drop the `@internal` annotations, which I could not combine.

MrKevinWeiss commented on drivers/modbus/Kconfig:
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2023-2024 Bas Stottelaar <basstottelaar@gmail.com>
```
We are no longer doing dependency modelling with kconfig... This can be dropped.

basilfx replied:
Right, so I can just remove the Kconfig file?

basilfx replied:
Removed the Kconfig file.

MrKevinWeiss reviewed (commented):
Really nice... I would like to see this in and would provide testing in the near future.

kfessel commented on drivers/include/modbus_rtu.h:
```diff
@@ -0,0 +1,193 @@
+#ifndef MODBUS_RTU_H
+#define MODBUS_RTU_H
+
+#include "periph/uart.h"
```
```suggestion
#include "periph/uart.h"
#include "periph/gpio.h"
```

basilfx replied:
Addressed (and sorted).

kfessel commented on drivers/include/modbus_rtu.h:
```diff
@@ -39,85 +40,35 @@ extern "C" {
+    const modbus_rtu_params_t* params;  /**< device parameters */
+    uint32_t timeout;       /**< amount of time (usec) to wait for a slave to
+                                 begin sending a */
+    uint32_t rx_timeout;    /**< time between two bytes before timeing out */
```
```suggestion
    uint32_t rx_timeout;    /**< timeout between bytes*/
```

basilfx replied:
Solved.

kfessel commented on drivers/include/modbus_rtu.h:
```diff
@@ -0,0 +1,144 @@
+    modbus_t dev;           /**< @ref modbus_t base class */
+    const modbus_rtu_params_t* params;  /**< device parameters */
+    uint32_t timeout;       /**< amount of time (usec) to wait for a slave to
+                                 begin sending a */
```
```suggestion
                                 begin sending*/
```

basilfx replied:
Solved.

basilfx commented:
I made some additional changes to the implementation.

* I removed message validation when using `modbus_rtu_send_request`. It is up to the caller to provide a valid message.
* I also removed message validation in `modbus_rtu_recv_request`. Validation should now happen in the callback, because it checks 'application level' properties.
* I added support for zero-copy read requests: in this case, the message will directly point to the internal buffer, which does not need copying.
* I added support for multi-slave implementations. Sounds like a lot, but I basically allow the request callback to handle any request message, so it is up to the callback to decide whether or not to send a response. This also allows for broadcasted requests (slave identifier == 0). It does require some additional checks in the callback, but I think its worth the flexibility.

kfessel commented on drivers/include/modbus_rtu.h:
```diff
@@ -0,0 +1,143 @@
+typedef struct {
+    modbus_t dev;                               /**< @ref modbus_t base class */
+    const modbus_rtu_params_t *params;          /**< device parameters */
+    uint32_t timeout;                           /**< amount of time (usec) to wait for a slave to
+                                                     begin sending */
+    uint32_t rx_timeout;                        /**< timeout between two bytes */
```
can u make them a little more descriptive 

like "inter_fame_usec_min" and "byte_usec_max" 

is that correct?

maybe its is more like frame_space_rx (1,5 chars ) , frame_space_ tx (3,5 chars)

please check me on this

basilfx replied:
I have moved `timeout` to the `params` section, and renamed `rx_timeout` to `byte_timeout`. 

rx_timeout is calculated during initialization, to save some computation in hot path (debatable).

Fields inside `modbus_rtu_t` are now marked as internal (with a comment).

derMihai commented on drivers/modbus/modbus_rtu.c:
```diff
@@ -0,0 +1,688 @@
+
+static void uart_cb(void *arg, uint8_t data)
+{
+    static msg_t msg;
```
This is never written to. Is this just a dummy message to wake-up the thread?

kfessel replied:
yes, it is

it's not uncommon in riot to do so, but other synchronization mechanisms might be a better choice
